cf-check/backup.c

```diff
@@ -8,14 +8,14 @@
 
 #if defined(__MINGW32__) || !defined(LMDB)
 
-int backup_main(int argc, char **argv)
+int backup_main(ARG_UNUSED int argc, ARG_UNUSED const char *const *const argv)
```

Comment:
I think it would be clearer to follow the language specifications in regards to modifiers here (modifier _after_ type), even though we are pretty inconsistent regarding `const` modifiers in other parts of the codebase.

I would prefer `ARG_UNUSED char const *const *const argv`, though it's probably not a big deal.

Reply:
@karlhto What do you mean by "following the language specifications"? [The ISO C11 standard](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf) calls these qualifiers, and does not specify whether you should put them before or after the first _type-specifier_ in a _type-name_, both are allowed by the syntax definition.

However, they do consistently use my preferred style, placing the const before the first _type-specifier_, in all the examples:

```C
const char *c_cp;
const void *c_vp;
extern int atoi(const char *);
```

```C
typedef int A[2][3];
const A a = {{4, 5, 6}, {7, 8, 9}}; // array of array of const int
int *pi;
const int *pci;
```

I would argue that `const char *const *` is the most readable way. The first `const` is unambiguous, intuitively, since there is nothing to the left it must apply to the right. You don't have to know the "const rule of thumb" to understand this. The second `const` is put together with the pointer asterisk (no space between) to indicate that it applies to this pointer. Most examples in the C standard follow this:

```C
const int *ptr_to_constant;
int *const constant_ptr;
```

(But there are also a couple of examples that use `* const`).

[glibc also seems to mostly follow the same convention.](https://github.com/bminor/glibc/blob/master/posix/execv.c#L23)

Reply:
Well, I might remember incorrectly, but I was pretty sure at least ANSI C required qualifiers to come after the type. IMO it might be slightly less ambiguous, because then you don't have the `const` qualifiers behaving differently for pointers and types, but keeping it consistent is a better idea anyway.

> I would argue that const char *const * is the most readable way. The first const is unambiguous, intuitively, since there is nothing to the left it must apply to the right. You don't have to know the "const rule of thumb" to understand this.

I agree with this. :)

Reply:
What I wrote above applies for both [ANSI C](https://www.pdf-archive.com/2014/10/02/ansi-iso-9899-1990-1/ansi-iso-9899-1990-1.pdf), as well as the K&R book. They don't say whether to put it before or after, the syntax definitions allow both, and they put it first in examples. Maybe you read it somewhere else (course material, blog post, etc.), but it doesn't seem to be part of any of the standards.

Reply:
Alright. Thanks for clarifying. :)